### PR TITLE
Provider update for service principal auth

### PR DESCRIPTION
**Pull request by mclacore**
Some proposed updates to the provider to fetch the data needed from the service principal so the app can auth to the docker registry in ACR.

* Removed `ApplicationID` because it was representing the object ID of the Enterprise Application
* Renamed `ServicePrincipalID` to `ServicePrincipalObjectID` to return the object ID of the application (service principal) for API requests
* Added `ServicePrincipalClientID` so that auth can be passed through on behalf of the service principal created

### CODE REVIEW

File: internal/provider/resource_application_identity.go [hunk 1]
```diff
@@ -79,11 +79,11 @@ var azureApplicationIdentityOutputs = tfsdk.Attribute{
 	Computed:    true,
 	Description: "Azure application and service principal configuration",
 	Attributes: tfsdk.SingleNestedAttributes(map[string]tfsdk.Attribute{
-		"application_id": {
+		"service_principal_object_id": {
```
Comment (WillBeebe) [inline internal/provider/resource_application_identity.go:82]: Maybe we keep `object` in golang vars but drop it here for a simpler interface.

File: internal/cloud/azure/application_identity.go [hunk 2]
```diff
@@ -78,13 +79,14 @@ func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityC
 	if spErr != nil {
 		return spErr
 	}
-	config.ServicePrincipalID = *sp.ID
-	config.ID = *sp.ID
+	config.ServicePrincipalObjectID = *sp.ID
+	config.ServicePrincipalClientID = *sp.AppId
```
Comment (WillBeebe) [inline internal/cloud/azure/application_identity.go:83]: Having this available is the actual "fix" for this PR. The rest is trying to make it easier to understand. 

When the app is created or fetched above, here's a snippet of the payload.

```
HTTP/1.1 200 OK
Content-type: application/json

{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications/$entity",
    "id": "03ef14b0-ca33-4840-8f4f-d6e91916010e",
    "deletedDateTime": null,
    "isFallbackPublicClient": null,
    "appId": "631a96bc-a705-4eda-9f99-fdaf9f54f6a2",
```

After that, when we create the service principal, that _also_ has an `appId`. It's my understanding that this is true:

+ app.ID == sp.ID
+ app.AppID == sp.AppID

Exposing the client ID (`*app.AppId`) gets us the right creds passed into the app's env-vars. 

```
app_settings = {
     DOCKER_REGISTRY_SERVER_URL      = data.azurerm_container_registry.main.login_server
     DOCKER_REGISTRY_SERVER_USERNAME = module.application.azure_application_identity.service_principal_client_id
     DOCKER_REGISTRY_SERVER_PASSWORD = module.application.azure_application_identity.service_principal_secret
   }
```

File: internal/cloud/azure/application_identity.go [hunk 3]
```diff
@@ -78,13 +79,14 @@ func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityC
...
+	config.ID = *sp.AppId
```
[the remark this answers is not in the document]
Comment (WillBeebe) [reply internal/cloud/azure/application_identity.go:84]: we tested having the AppId here but using the service principal id as the identity is definitely what we want.

**Review by WillBeebe — CHANGES_REQUESTED** [review]
### additional context

When we did our first pass the app was created, deployed, pulled the Docker image, etc.. We worked through the Terraform for the Azure function and missed one debug line where we hard-coded an id.


### Issue / motivation

First, wow, Azure gets more heady the more I learn! Michael did a great job of explaining the app, app id, object service principal mind mess that's Azure. He was making changes to these files and building the provider locally as we debugged / worked through this. 

We ended up adding `object_id` to some variables to help clarify _exactly_ what's going on here. `app.ID` is actually the `object_id` for the app, it has another field that's `app.AppId`, which is super fun. Also, the app's `object_id` is the service principal id (like the comment says in the code).

This, yes this. The `object id` property of the _application_ is the `service_principal_object_id`. 🤯 



- [changed] ApplicationID => ServicePrincipalObjectID
- [added] ServicePrincipalClientID

If `Object` is too much, I could see dropping that back to `ServicePrincipalID` but it was helpful today.

File: internal/cloud/azure/application_identity.go (hunk 2 above)
Comment (chrisghill) [reply internal/cloud/azure/application_identity.go:83]: I think you're right that app.AppId == sp.AppId, but I don't think it's true that app.Id == sp.Id.

File: internal/cloud/azure/application_identity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisghill) [reply internal/cloud/azure/application_identity.go:84]: Be careful with this. The `config.Id` is intended to be able to be used in the `application_permission` resource as the `application_identity_id` field.

In AWS its the role ARN, in GCP its the service account e-mail (both are the entities that permissions can be bound to). It took a couple attempts for me to make sure that the value `config.ID` for an Azure `application_identity` was the ID of what `application_permission` would need to assign permissions. `sp.ID` was what I found that worked. Though, I never tested the actual permissions using a service principal to see if it actually worked.

If this is tested and working, then fantastic. But if you haven't verified this ^ works we should do so before we merge.

File: internal/provider/resource_application_identity.go (hunk 1 above)
Comment (chrisghill) [reply internal/provider/resource_application_identity.go:82]: Yeah, when I named these I tried to keep them similar to the counterparts I identified in the official `azurerm` provider. The idea being that it will be a similar interface for anybody that already uses azure for managing terraform.

**Review by chrisghill — COMMENTED** [review]
Looks like we found a new Golang developer!

I like the changes, but I left some notes about some concerns.

File: internal/cloud/azure/application_identity.go [hunk 4]
```diff
@@ -106,14 +108,14 @@ func UpdateApplicationIdentity(ctx context.Context, config *ApplicationIdentityC
...
-	_, appErr := appClient.Delete(ctx, config.ApplicationID)
+	_, appErr := appClient.Delete(ctx, config.ApplicationObjectID)
 	if appErr != nil {
 		return appErr
 	}
 
 	config.ID = ""
-	config.ApplicationID = ""
-	config.ServicePrincipalID = ""
+	config.ServicePrincipalObjectID = ""
```
[the remark this answers is not in the document]
Comment (mclacore) [reply internal/cloud/azure/application_identity.go:117]: Will do!

File: internal/provider/resource_application_identity.go (hunk 1 above)
Comment (mclacore) [reply internal/provider/resource_application_identity.go:82]: The discovery we made yesterday though is the `Application.go` package in the Hamilton repo is using **Azure Enterprise Applications**, which is different than **Azure App Registration** (which is where service principals are configured). In some edge cases, the appID of the service principal is not the appID of the enterprise application. I made the change to add clarity of what the actual value is coming through as, because the `service_principal_object_id` is for API requests, but the service principal's appID is used for auth/roles/etc.

Basically, we were slamming our heads against the wall yesterday trying to figure out why the appID of the application, which should be giving us the ability to auth to the ACR, was failing.

File: internal/cloud/azure/application_identity.go (hunk 2 above)
Comment (mclacore) [reply internal/cloud/azure/application_identity.go:83]: > I could bind roles to the sp.Id, but not to the app.Id.

We were binding roles using the `sp.Id`, but using the `sp.Id` as the principal user to auth to ACR was failing, because it needed the principal ID of the service principal, which via the API is generated via `sp.AppId`

File: internal/cloud/azure/application_identity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (WillBeebe) [reply internal/cloud/azure/application_identity.go:84]: You're spot-on here Chris! I led Michael down this path when we were trying to figure things out but `sp.ID` is what we want here. 👍

File: internal/provider/resource_application_identity.go (hunk 1 above)
Comment (WillBeebe) [reply internal/provider/resource_application_identity.go:82]: Since this is the "exposed" interface that we'd interact w/ in Terraform. Let's make this name `service_principal_id`

**Comment by WillBeebe** [comment]
I think there are only two changes before this PR is gtg. Mentioned above ^

File: internal/provider/resource_application_identity.go (hunk 1 above)
Comment (mclacore) [reply internal/provider/resource_application_identity.go:82]: If this is the exposed object we interact with, I think it would be best to make it clear about what exactly is being passed since `id` is vague in Azure's terms.

Comment (WillBeebe) [reply internal/provider/resource_application_identity.go:82]: Personally, I agree since Azure is hard! But Chris has a good point here. In Terraform for Azure, I haven't seen `service_principal_object_id` but I have seen `service_principal_id`. He's saying that we should match what's used in the Azure provider and other Azure TF so when people see this value, they know what it means, even if object id is technically clearer.

File: internal/cloud/azure/application_identity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mclacore) [reply internal/cloud/azure/application_identity.go:84]: I agree here, I made the change and tested it

[the remark this answers is not in the document]
Comment (WillBeebe) [reply internal/cloud/azure/application_identity.go:84]: Rockstar, thank you!

File: internal/provider/resource_application_identity.go (hunk 1 above)
Comment (mclacore) [reply internal/provider/resource_application_identity.go]: Change has been made

File: internal/cloud/azure/application_identity.go [hunk 5]
```diff
@@ -79,12 +80,13 @@ func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityC
...
 	config.ServicePrincipalID = *sp.ID
+	config.ServicePrincipalClientID = *sp.AppId
 	config.ID = *sp.ID
 
 	// fetch the service principal to make sure it exists
-	_, _, spCheckErr := spClient.Get(ctx, config.ServicePrincipalID, odata.Query{})
+	_, _, spCheckErr := spClient.Get(ctx, config.ID, odata.Query{})
 	if spCheckErr != nil {
-		return fmt.Errorf("error retrieving Service Principal with ID %v: %w", config.ServicePrincipalID, spCheckErr)
+		return fmt.Errorf("error retrieving Service Principal with ID %v: %w", config.ServicePrincipalClientID, spCheckErr)
```
Comment (chrisghill) [inline internal/cloud/azure/application_identity.go:89]: Why are we swapping the types here? We check if it exists using the `config.ID` then if it fails in the error message we're outputing a different value (`config.ServicePrincipalClientId`)

File: internal/mdxc/application_identity.go [hunk 6]
```diff
@@ -28,9 +28,9 @@ type AWSApplicationIdentityOutputData struct {
 	IAMRoleARN types.String `tfsdk:"iam_role_arn"`
 }
 type AzureApplicationIdentityOutputData struct {
-	ApplicationID          types.String `tfsdk:"application_id"`
```
Comment (chrisghill) [inline internal/mdxc/application_identity.go:31]: Why are we removing `ApplicationId`? Is it not useful?  I believe the only way to look up the Application (for `Read` or `Update`) is via the application ID.  Here is the query: https://github.com/massdriver-cloud/terraform-provider-mdxc/pull/28/files?diff=split&w=1#diff-1c8618fd6e267412ff00f04541e24eaca469da6aea403c53bd94d295201aa585R71

If we need to **add** ServicePrincipalClientID thats fine, but we shouldn't remove ApplicationID

Comment (mclacore) [reply internal/mdxc/application_identity.go:31]: I agree `ApplicationID` is still useful. Newer commits brought it back, but I missed this line. Next commit will revert this change here.

File: internal/cloud/azure/application_identity.go (hunk 5 above)
Comment (mclacore) [reply internal/cloud/azure/application_identity.go:89]: I recall this was a change made when still trying to discover what the actual values contained. Next commit will change this value to `config.ID`

File: internal/mdxc/application_identity.go (hunk 6 above)
Comment (chrisghill) [reply internal/mdxc/application_identity.go:31]: We need it in a few places:
https://github.com/massdriver-cloud/terraform-provider-mdxc/pull/28/files#diff-08d112e976e2479674666bd6acea8430fd217225c39ce5186b3e1a4788d52ad1L82
https://github.com/massdriver-cloud/terraform-provider-mdxc/pull/28/files#diff-08d112e976e2479674666bd6acea8430fd217225c39ce5186b3e1a4788d52ad1L48
https://github.com/massdriver-cloud/terraform-provider-mdxc/pull/28/files#diff-19a6e506204b4de650e1289aa75ded86e795860b5da91100431d3c685fc8fe1dL160

**Review by chrisghill — APPROVED** [review]
